maven/lib/dependabot/maven/version.rb

```diff
@@ -11,181 +12,10 @@
-        version = version.gsub(/(\d)([A-Za-z])/, '\1-\2')
-        version = version.gsub(/([A-Za-z])(\d)/, '\1-\2')
-
-        # Replace empty tokens with 0
-        version = version.gsub(/([\.\-])([\.\-])/, '\10\2')
-        version = version.gsub(/^([\.\-])/, '0\1')
-        version.gsub(/([\.\-])$/, '\10')
-      end
-
-      def trim_version(version)
-        version.split("-").filter_map do |v|
-          parts = v.split(".")
-          parts = parts[0..-2] while NULL_VALUES.include?(parts&.last)
-          parts&.join(".")
-        end.reject(&:empty?).join("-")
-      end
-
-      def convert_dates(version, other_version)
-        default = [version, other_version]
-        return default unless version.match?(/^\d{4}-?\d{2}-?\d{2}$/)
-        return default unless other_version.match?(/^\d{4}-?\d{2}-?\d{2}$/)
-
-        [version.delete("-"), other_version.delete("-")]
-      end
-
-      def split_into_prefixed_tokens(version)
-        ".#{version}".split(/(?=[\-\.\+])/)
-      end
-
-      def pad_for_comparison(prefixed_tokens, other_prefixed_tokens)
-        prefixed_tokens = prefixed_tokens.dup
-        other_prefixed_tokens = other_prefixed_tokens.dup
-
-        longest = [prefixed_tokens, other_prefixed_tokens].max_by(&:count)
-        shortest = [prefixed_tokens, other_prefixed_tokens].min_by(&:count)
-
-        longest.count.times do |index|
-          next unless shortest[index].nil?
-
-          shortest[index] = longest[index].start_with?(".") ? ".0" : "-"
-        end
-
-        [prefixed_tokens, other_prefixed_tokens]
-      end
-
-      def compare_prefixed_token(prefix:, token:, other_prefix:, other_token:)
-        token_type = token.match?(/^\d+$/) ? :number : :qualifier
-        other_token_type = other_token.match?(/^\d+$/) ? :number : :qualifier
-
-        hierarchy = PREFIXED_TOKEN_HIERARCHY.fetch(prefix).fetch(token_type)
-        other_hierarchy =
-          PREFIXED_TOKEN_HIERARCHY.fetch(other_prefix).fetch(other_token_type)
-
-        hierarchy_comparison = hierarchy <=> other_hierarchy
-        return hierarchy_comparison unless hierarchy_comparison.zero?
-
-        compare_token(token: token, other_token: other_token)
-      end
-
-      def compare_token(token:, other_token:)
-        if (token_hierarchy = NAMED_QUALIFIERS_HIERARCHY[token])
-          return -1 unless NAMED_QUALIFIERS_HIERARCHY[other_token]
-
-          return token_hierarchy <=> NAMED_QUALIFIERS_HIERARCHY[other_token]
-        end
-
-        return 1 if NAMED_QUALIFIERS_HIERARCHY[other_token]
-
-        if token.match?(/\A\d+\z/) && other_token.match?(/\A\d+\z/)
-          token = token.to_i
-          other_token = other_token.to_i
-        end
-
-        token <=> other_token
+    class Version < Dependabot::NewVersion
+      sig { override.overridable.params(version: VersionParameter, _version_string: T.nilable(String)).void }
+      def initialize(version, _version_string = nil)
+        super(version.to_s.tr("_", "-"), version.to_s)
```

Comment:
do we have tests coverage?

Reply:
Also, any reason why we are doing the replace?

Reply:
what's the `_version_string` parameter used for?

Reply:
> Also, any reason why we are doing the replace?

The spec for the ruby superclass, `Gem::Version` works with dashes but not underscores. So without replacing we'd end up with the malformed version error.

Reply:
Ahh I see; do we have tests already covering this specific behaviour?

Reply:
> Also this comment "what's the _version_string parameter used for?"

It's not used but sorbet insisted that I add it. The `initialize` method in `new_version.rb` has an additional optional parameter to store the original version string and because I'm overriding it in here sorbet recommended doing it this way.

Reply:
Should it be a private variable for `new_version` then?

Reply:
I'm looking into it. If sorbet doesn't throw inexplicable errors I'll make it private

Reply:
@abdulapopoola sorbet still complains even if I make the variable private. Bacause `initialize` in new_version.rb accepts 2 keyword arguments, it expects every subclass of new_version to do the same.

Reply:
That's the question; why does new_version need that parameter if it is only used privately?